### PR TITLE
Parity trie solidity implementation (2/3)

### LINKED ISSUE #151: Parity scale codec - solidity implementation
https://github.com/paritytech/parity-scale-codec
https://github.com/paritytech/trie/blob/master/trie-db/src/node_codec.rs#L34

This codec is needed to be implemented in solidity for the deserialization of node data in the substrate state proof.
This is a required component for Parity trie implementation #98 

CC: @seunlanlege please let me know when we can discuss the scope of this implementation.

### PR DESCRIPTION
**Pull request by nadeemb53**
Closes #151 

PR split #152

### CODE REVIEW

File: contracts/ethereum/src/core/node-codec/NodeBuilder.sol [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+contract NodeBuilder is ICodec, ITrie {
+    NodeHandle[] emptyChildren;
+    NodeHandle emptyChild;
+    Slice emptySlice;
+    Value emptyValue;
+
+    uint8 constant NIBBLE_LENGTH = 16;
+
+    function buildNode(NodePlanStruct memory nodePlan, uint8[] calldata data)
```
Comment (blasrodri) [inline contracts/ethereum/src/core/node-codec/NodeBuilder.sol:15]: can you please add more docs on the functions?

File: contracts/ethereum/src/core/node-codec/NodeBuilder.sol [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+    function buildValue(ValuePlan memory valuePlan, uint8[] calldata data)
+        internal
+        pure
+        returns (Value memory)
+    {
+        if (valuePlan.isInline) {
+            return Value(true, data[valuePlan.start:valuePlan.end]);
+        }
+        return Value(false, data[valuePlan.start:valuePlan.end]);
```
Comment (blasrodri) [inline contracts/ethereum/src/core/node-codec/NodeBuilder.sol:98]:
```suggestion
        return Value(valuePlan.isInline, data[valuePlan.start:valuePlan.end]);
```

File: contracts/ethereum/src/core/node-codec/NodeBuilder.sol [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    function buildNodeHandle(
+        NodeHandlePlan memory nodeHandlePlan,
+        uint8[] calldata data
+    ) internal pure returns (NodeHandle memory) {
+        if (nodeHandlePlan.isHash) {
+            return
+                NodeHandle(true, data[nodeHandlePlan.start:nodeHandlePlan.end]);
```
Comment (blasrodri) [inline contracts/ethereum/src/core/node-codec/NodeBuilder.sol:107]: same suggestion than on `buildValue`?

File: contracts/ethereum/src/core/node-codec/NodeBuilder.sol (hunk 1 above)
Comment (nadeemb53) [reply contracts/ethereum/src/core/node-codec/NodeBuilder.sol:23]: done

File: contracts/ethereum/src/core/trie-db/LookUp.sol [hunk 4]
```diff
@@ -235,4 +218,9 @@ contract LookUp is ITrie {
         }
```
[the remark this answers is not in the document]
Comment (nadeemb53) [reply contracts/ethereum/src/core/trie-db/LookUp.sol]: Yeah, I've removed the implementation for now. It needs to be implemented differently which I'll do in the next PR. For context, it doesn't have to be bytes to bytes32 conversion but a uint8 array to bytes conversion depending upon the hasher (blake2b hash is 64 bytes long).

**Review by blasrodri — APPROVED** [review]

**Pull request merged**